### PR TITLE
crypto: add constant time equal and rand gen

### PR DESCRIPTION
**Pull request by stlaz**
Currently used in OAuth server and its operator, these are generally useful.

### CODE REVIEW

**Comment by enj** [comment]
/lgtm

Small but not code that I want repeated everywhere.

**Comment by enj** [comment]
Note that this is also used by the console operator to generate its OAuth client secret.

File: pkg/crypto/compare.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package crypto
+
+import "crypto/subtle"
+
+func IsEqualConstantTime(s1, s2 string) bool {
```
Comment (sttts) [inline pkg/crypto/compare.go]: docs. What is this?

Comment (deads2k) [inline pkg/crypto/compare.go]: nak, no wrappers on one liners

File: pkg/crypto/random.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+)
+
+// RandomBits returns a random byte slice with at least the requested bits of entropy.
+// Callers should avoid using a value less than 256 unless they have a very good reason.
+func RandomBits(bits int) []byte {
+	size := bits / 8
```
Comment (sttts) [inline pkg/crypto/random.go]: `size := (bits + 7) / 8`

File: pkg/crypto/random.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package crypto
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+)
+
+// RandomBits returns a random byte slice with at least the requested bits of entropy.
+// Callers should avoid using a value less than 256 unless they have a very good reason.
+func RandomBits(bits int) []byte {
```
Comment (deads2k) [inline pkg/crypto/random.go]: nak, never used outside this. excessive flexibility

File: pkg/crypto/random.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	if _, err := rand.Read(b); err != nil {
+		panic(err) // rand should never fail
+	}
+	return b
+}
+
+// RandomBitsString returns a random string with at least the requested bits of entropy.
+// It uses RawURLEncoding to ensure we do not get / characters or trailing ='s.
+func RandomBitsString(bits int) string {
+	return base64.RawURLEncoding.EncodeToString(RandomBits(bits))
```
Comment (deads2k) [inline pkg/crypto/random.go]: nak, no one liners

File: pkg/crypto/random.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// Random256BitsString is a convenience function for calling RandomBitsString(256).
+// Callers that need a random string should use this function unless they have a
+// very good reason to need a different amount of entropy.
+func Random256BitsString() string {
```
Comment (deads2k) [inline pkg/crypto/random.go]: leaving this function which becomes four lines long.  

nak, copy.

**Comment by deads2k** [comment]
I think you can get a away with one liners (just embed them, wrappers aren't helpful) and a four line function (at most)  in two spots.

/close

**Comment by deads2k** [comment]
If you can find three callers in different binaries, I can live with Random256 and Random512.

**Comment by deads2k** [comment]
/hold

for evidence

**Comment by enj** [comment]
> If you can find three callers in different binaries, I can live with Random256 and Random512.

1. OAuth server
1. Authn operator
1. Console operator

I am sure I can find others.

**Comment by deads2k** [comment]
> > If you can find three callers in different binaries, I can live with Random256 and Random512.
> 
> 1. OAuth server
> 2. Authn operator
> 3. Console operator
> 
> I am sure I can find others.

if you're moving it here to be useful you should find them all.

> > If you can find three callers in different binaries, I can live with Random256 and Random512.
> 
> 1. OAuth server
> 2. Authn operator
> 3. Console operator
> 
> I am sure I can find others.

Then do so.

**Comment by deads2k** [comment]
@tnozicka verify-deps again

**Comment by enj** [comment]
1. oc
1. extended tests (I do not really count this though)

GitHub search is lacking in this regard.

**Comment by stlaz** [comment]
Removed the constant-time-comparison, kept only the `RandomBitsString*` functions, while keeping their base still exporrted

**Comment by enj** [comment]
/unassign
/uncc

@stlaz @sttts @mfojtik

**Pull request closed without merging**